### DISCUSSION TITLE
TX/RX side selection via configuration

### DISCUSSION THREAD

**Discussion by PhilippMolitor** [ExpressLRS/ExpressLRS#2357] [Ideas]
This idea proposes to make RX/TX side selection available from configuration, not at compile time.

# Why would someone want this?

The reason I came up with this idea is Airport - telemetry output power is limited to the RX hardware. Usually, that is not well documented and maxes out at 250mW in best cases. For more serious projects (Think something like a **big** plane or quad for long range or extended autonomy) it would be a great improvement to use a cheap 1W TX module with cooling on the Air side of Airport.

Also, this would enable to use RXes as cheap TX hardware for any kind of electronics/hobby project.

If you have a different use case from that on why you would want to switch out hardware purpose and software TX/RX side, please let me know why - maybe I overlooked an interesting use case.

# Current implementation

Right now, each target gets a `define` at compile time to determine if the hardware is a transmitter or receiver
(see [common.ini](https://github.com/ExpressLRS/ExpressLRS/blob/567bc8c4fbb574330af70f9162586a43e2bbaaca/src/targets/common.ini#L16-L17)).

# Desired implementation

Move the choice of TX/RX to software configuration. For best usability, it should be a toggle in the web UI, so at any point in time the functionality can be switched around.

# Pitfalls

To my knowledge the UART on the TX side is half-duplex. Directly wiring that port to an FC could introduce issues. Maybe an unused Pin can be added into the target definitions to enable full-duplex UART communication on TXes if desired.

**Comment by pkendall64** [comment]
Its a lot more involved than what you made it look like in your "Current Implementation" section above!
The main code for the TX and RX are completely different files.

**Reply by PhilippMolitor** [reply]
Thanks for letting me know! So it seems like compiling the appropriate edition is still necessary. Maybe there could be a switch for the flashing configuration UI then?

**Comment by sdellava** [comment]
I've a use case for this feature request I've posted here: https://github.com/ExpressLRS/ExpressLRS/discussions/2340

Basically, due to the fact that the RX and TX boards, from the radio physical layer point of view, are identical, there is not reason do deny to install a TX fw on a RX board and vice versa.

In addition, for other weired use cases, I'd suggest to include the option to disable the telemetry on RX side. ELRS protocol has the unique feature to bind on a passphrase, so it is possible to drive many RX at the same time. Each of them can use specifics channels. This scenario can be very usefull in many use cases where, due to the low cost of the ELRS boards, a set of RXs can be used to control many servo/motor instead of use one big RX.  In this scenario only the master RX have to transmit the telemetry while the other have to be silent. We can assume that all rx are in the same radio conditions so it is not requred to have each telemetry feed from each of them.

**Reply by pkendall64** [reply]
> In addition, for other weired use cases, I'd suggest to include the option to disable the telemetry on RX side. ELRS protocol has the unique feature to bind on a passphrase, so it is possible to drive many RX at the same time. Each of them can use specifics channels. This scenario can be very usefull in many use cases where, due to the low cost of the ELRS boards, a set of RXs can be used to control many servo/motor instead of use one big RX. In this scenario only the master RX have to transmit the telemetry while the other have to be silent. We can assume that all rx are in the same radio conditions so it is not requred to have each telemetry feed from each of them.

This already exists

**Comment by sdellava** [comment]
Can you please explain what exists?

Il mar 22 ago 2023, 21:03 Paul Kendall ***@***.***> ha
scritto:

**Reply by pkendall64** [reply]
The ability do disable telemetry output on receivers so you can have multiple receivers connected simultaneously to a transmitter.
Theres an option in the receiver web UI to force disable the telemetry.

**Comment by sdellava** [comment]
Thanks, I'll check it out.

Il mar 22 ago 2023, 21:09 Paul Kendall ***@***.***> ha
scritto:
